### ISSUE TITLE
Tax calculation for "Custom-Rates" plugin leads to an infinite looped code

### ISSUE THREAD

**Issue by rohit-elevar** [reactioncommerce/reaction#4620]
## Issue Description
1. The **"taxes/calculate"** method from `/imports/plugins/core/taxes/server/methods/methods.js` is called after the address is entered by the user.
2. This in turn calculated the tax values and calls **"taxes/setRateByShopAndItem"** from the same file to actually update the respective record in the Cart collection. This in turn emits an afterCartUpdate event **appEvents.emit("afterCartUpdate"...**
3. The listener **appEvents.on("afterCartUpdate"...** in `/imports/plugins/core/discounts/server/hooks/cart.js` checks for discounts and then runs the hook: **afterCartUpdateCalculateTaxes**
4. The listener for **"afterCartUpdateCalculateTaxes"** in `/imports/plugins/core/taxes/server/hooks/taxes.js` calls the **"taxes/calculate"** method again.

**This is causing the code to loop.**

## Steps to Reproduce
1. Add debugging statements to 3 files [ Logger.info() / console.log() ]
    1. /imports/plugins/core/taxes/server/methods/methods.js in the **taxes/setRateByShopAndItem** method
    2. /imports/plugins/core/discounts/server/hooks/cart.js in the **afterCartUpdate** listener
    3. /imports/plugins/core/taxes/server/hooks/taxes.js in the **afterCartUpdateCalculateTaxes** listener
2. Login as admin
3. Confirm that the default reaction product "Is taxable" option is enabled for all variants
4. Change currency in Localization to *INR*, and disable all other currencies [probably optional, but I did this]
5. Enable Custom Rates in the Tax section of dashboard.
6. Create a new tax rate with *Country: India* and *Rate: 12*
7. Enable Flat Rate under Shipping. And also enable "Free" Shipping option and click Save Changes.
8. Log out
9. Add the product to cart
10. Checkout
11. Continue as guest, and enter email
12. Enter the address field. Enter any address but choose country as *IN* (India)
13. Once you submit the address form, the shipping option will not show up
14. Check the console for the logs you added in Step 1

### Versions
Node: 8.11.4
NPM: 5.6.0
Meteor Node: 8.11.3
Meteor NPM: 5.10.0
Reaction CLI: 0.29.0
Reaction: 1.16.0
Reaction branch: master
Docker: 18.06.1-ce

**Comment by spencern** [comment]
Hi @rohit-elevar 
Thanks so much for the very detailed bug report and reproduction steps, it's clear you've put a lot of time into helping us resolve this issue and I appreciate that. ❤️ 

A few questions:
1. It seems that this issue requires changing core code as seen in step 2? Does this issue appear if the code in step 2 is left unchanged?
2. I'm not fully following why you're doing that - are you having issues with taxes not working if the code is left as is?

**Comment by rohit-elevar** [comment]
Hi @spencern 
I'm new to both reaction and git, and this was the first issue I had created, so I thought it would be better to err on the side of clarity :)

1. Yes the issue still occurs even if the step 2 is left unchanged. I had made that change in my code base, so I added it in the steps here too. Should it be removed and made into a new bug report?
2. Yes, the custom tax rates doesn't work unless the change in step 2 is made. The taxes are always zero.

**Comment by spencern** [comment]
@rohit-elevar 
Thanks for clearing that up. If you'd be willing to file a separate issue for the taxes not working I'd appreciate that. In general, bugs reported that are caused by a change to the core code are seen as an issue with the change to our core code and not with our system.

**Comment by rohit-elevar** [comment]
I've made the required edits to this bug report. I'll create another one for the taxes not working

**Comment by spencern** [comment]
@rohit-elevar thank you, we'll take a look.

**Comment by aldeed** [comment]
FWIW, this code is rewritten in 2.0 and I don't believe will have this issue anymore. We may have to also fix in a different way in 1.x. I had already been thinking we should pass an extra `name` argument when emitting the events, and if the name is the same as current function's name, don't run again.

In 2.0, it does not update or emit if it doesn't have any changes, so at most it would run twice.

It might also be possible to fix `appEvents` more transparently, by tracking that we're inside another `emit` and don't run the same function that is emitting (skip if `currentlyRunningFunc === func`)

**Comment by spencern** [comment]
@aldeed @rohit-elevar If this has been rewritten for 2.0, we'll look at a PR to resolve it on the 1.x branch but it will probably not be a focus of the core team in the near term.

**Comment by rohit-elevar** [comment]
@spencern @aldeed I understand that this won't be a primary focus. But since I cannot move forward without resolving this I just a couple of questions:
1. When is the 2.0 release tentatively scheduled for
2. Am I correct in assuming the the way to temporarily resolve this would be to check if the cart object has changes from the previous time the event was called? If yes, any guidance/direction on how I can proceed to do it?

**Comment by aldeed** [comment]
@rohit-elevar 2.0 is scheduled for this Thursday.

This simplest fix is probably to update `setRateByShopAndItem` to this:

```js
const { cartTaxData, cartTaxRate, itemsWithTax, taxRatesByShop } = options;

const cart = Cart.findOne({ _id: cartId });

let result;
if (
  !_.isEqual(cart.taxes, cartTaxData) ||
  !_.isEqual(cart.tax, cartTaxRate) ||
  !_.isEqual(cart.items, itemsWithTax) ||
  !_.isEqual(cart.taxRatesByShop, taxRatesByShop)
) {
  result = Cart.update({ _id: cartId }, {
    $set: {
      taxes: cartTaxData,
      tax: cartTaxRate,
      items: itemsWithTax,
      taxRatesByShop
    }
  });

  const updatedCart = Cart.findOne({ _id: cartId });
  Promise.await(appEvents.emit("afterCartUpdate", cartId, updatedCart));
}
```

**Comment by rohit-elevar** [comment]
Thanks a lot for that @aldeed!!

**Comment by aldeed** [comment]
Added this fix to 1.17.0 as well.
